### PR TITLE
Bind fastify instance in hooks

### PR DESCRIPTION
**Pull request by StarpTech**

### CODE REVIEW

**Comment by StarpTech** [comment]
```
```

**Comment by StarpTech** [comment]
I observed a strange behavior during benchmarking. 

1. Checkout master run autocannon ~ 420-450k
2. Checkout #120_context_hook run autocannon ~ 415-435k
3. Restart ./examples/simple.js multiple times I can reach only ~ 340-357k

Anybody why? Does V8 any caching magic even when we have the same file with different content?

**Comment by StarpTech** [comment]
In Node.js 6.11.5 I can't measure a performance regression I think the reason for the different has something to do with Ignition + TurboFan execution pipeline released in Node.Js 8.3

**#120_context_hook**
```
Running 10s test @ http://localhost:3000/
100 connections with 10 pipelining factor

Stat         Avg     Stdev   Max
Latency (ms) 2.5     6.1     210
Req/Sec      35330.4 5198.13 38655
Bytes/Sec    5.24 MB 772 kB  5.77 MB

353k requests in 10s, 52.6 MB read
```

**master**
```
Running 10s test @ http://localhost:3000/
100 connections with 10 pipelining factor

Stat         Avg      Stdev   Max
Latency (ms) 2.46     7.28    286
Req/Sec      35752.81 3386.18 37919
Bytes/Sec    5.33 MB  503 kB  5.77 MB

357k requests in 10s, 53.3 MB read
```

**Comment by mcollina** [comment]
can we bind the functions when they are added, rather than using `call`  at Runtime?

**Comment by StarpTech** [comment]
Good question, I will investigate.

**Comment by allevo** [comment]
Can you re-run the benchmark?

**Comment by StarpTech** [comment]
**master**
```
Running 10s test @ http://localhost:3000/
100 connections with 10 pipelining factor

Stat         Avg     Stdev   Max
Latency (ms) 1.94    5.75    235
Req/Sec      44496   3768.38 47231
Bytes/Sec    6.63 MB 560 kB  7.08 MB

445k requests in 10s, 66.3 MB read
```

**#120_context_hook**
```
Running 10s test @ http://localhost:3000/
100 connections with 10 pipelining factor

Stat         Avg     Stdev   Max
Latency (ms) 1.91    5.94    265
Req/Sec      44918.4 4044.65 47423
Bytes/Sec    6.63 MB 572 kB  7.08 MB

449k requests in 10s, 66.9 MB read
```

File: test/hooks.test.js [hunk 1]
```diff
@@ -145,10 +145,12 @@ test('onRequest hook should support encapsulation / 2', t => {
 })
 
 test('onRequest hook should support encapsulation / 3', t => {
-  t.plan(13)
+  t.plan(17)
   const fastify = Fastify()
+  fastify.decorate('hello', 'world')
 
-  fastify.addHook('onRequest', (req, res, next) => {
+  fastify.addHook('onRequest', function (req, res, next) {
```
Comment (delvedor) [inline test/hooks.test.js]: Can you also add a decorator *after* the hook registration?

Comment (StarpTech) [reply test/hooks.test.js]: What's the reason?

Comment (delvedor) [reply test/hooks.test.js]: To be 100% sure that we are not missing the context reference :)

Comment (StarpTech) [reply test/hooks.test.js]: I mean what does it verify. I don't get it. Can you make an example?

Comment (delvedor) [reply test/hooks.test.js]: Unlike the routes the hooks are registered synchronously.
We had some troubles with the async declaration of the routes, so I want to add a test now for the hooks, in this way if in the future we make thew hook api async, we can detect very quickly weird behaviours.

Comment (StarpTech) [reply test/hooks.test.js]: Check

**Comment by StarpTech** [comment]
Done, please check for typos :)

File: docs/Hooks.md [hunk 2]
```diff
@@ -131,7 +131,15 @@ fastify.addHook('onClose', (instance, done) => {
 ```
 <a name="scope"></a>
 ### Scope
-Except for `'onClose'` all the hooks are encapsulated this means that you can decide where your hooks should run by using `register` as explained in the [plugins guide](https://github.com/fastify/fastify/blob/master/docs/Plugins-Guide.md).
+Except for `'onClose'` all the hooks are encapsulated this means that you can decide where your hooks should run by using `register` as explained in the [plugins guide](https://github.com/fastify/fastify/blob/master/docs/Plugins-Guide.md). If you pass a function, that function is bind to the right fastify context and from there you have full access to the fastify api.
```
Comment (jsumners) [inline docs/Hooks.md]: "bind to" => "bound to"

Comment (StarpTech) [reply docs/Hooks.md]: thanks, fixed.

File: docs/Hooks.md [hunk 3]
```diff
@@ -131,7 +131,15 @@ fastify.addHook('onClose', (instance, done) => {
 ```
 <a name="scope"></a>
 ### Scope
-Except for `'onClose'` all the hooks are encapsulated this means that you can decide where your hooks should run by using `register` as explained in the [plugins guide](https://github.com/fastify/fastify/blob/master/docs/Plugins-Guide.md).
+Except for `'onClose'` all the hooks are encapsulated this means that you can decide where your hooks should run by using `register` as explained in the [plugins guide](https://github.com/fastify/fastify/blob/master/docs/Plugins-Guide.md). If you pass a function, that function is bound to the right Fastify context and from there you have full access to the Fastify api.
+
+```js
+fastify.addHook('onRequest', (req, res, next) => {
```
Comment (delvedor) [inline docs/Hooks.md]: This is not correct, you should not use an arrow function to access the correct this :D

Comment (StarpTech) [reply docs/Hooks.md]: :D

Comment (StarpTech) [reply docs/Hooks.md]: Done

**Comment by delvedor** [comment]
@StarpTech for the future I'd like to wait that both Travis and AppVeyor are green before merge.
It is useless have two CI and use just one :)

**Comment by StarpTech** [comment]
I agree but I verified Windows locally and Appveyor queue can take sometimes hours. Additionally, this change was very small and I could exclude node incompatibilities.

**Pull request merged**